### ISSUE TITLE
Zazu retaining focus after search window is closed.

### ISSUE THREAD

**Issue by PAB-me-uk** [bayleeadamoss/zazu#185]
### Expected Behavior

When Zazu is hidden the focus should be returned to the previous application.

### Actual Behavior

When Zazu is hidden Zazu retains the focus even though the Zazu window is hidden.

### Steps to reproduce the behavior

* Press toggle key to show Zazu.

* Press escape or toggle key again, or select an action that doesn't launch a new window (e.g. clipboard plugin)

* The Zazu search window is hidden

* The focus is not returned to the previous application.

* Pressing ctrl+F4 (Windows close application shortcut) closes the Zazu application removing the icon from the task bar which proves that Zazu still had the focus.

## Versions

Zazu: 0.3.2
OS: Windows 8.1 (probably all)
GIT: 2.11

## Notes

This issue is particularily problematic for the clipboard plugin as after selecting a clip entry I have to manually click on the applications window to get the focus, then press ctrl+v to paste.

I think this may actually be an issue with Electron see https://github.com/electron/electron/issues/3472

**Comment by bayleedev** [comment]
I believe this is a restriction of electron, as you suggested. On Mac there is a `app.hide()` api we call [here](https://github.com/tinytacoteam/zazu/blob/master/app/background.js#L116), but I'm not sure how to do this on windows/linux yet, but would love to get this fixed.

**Comment by arieljannai** [comment]
@PAB-me-uk 
The issue you linked to is closed, so I guess it's not something that's going to change from electron's side?

**Comment by PAB-me-uk** [comment]
I have requested that https://github.com/electron/electron/issues/3472 is reopened as it has not been fixed.

**Comment by arieljannai** [comment]
I'm trying to create an [AHK](https://autohotkey.com/) script as a workaround

**Comment by PAB-me-uk** [comment]
One workaround would be to use the windows api functions as follows:

When activated by hot key use [GetFocus](https://msdn.microsoft.com/en-us/library/windows/desktop/ms646294(v=vs.85).aspx) to get the hwnd of the current app.

When hiding window use [SetFocus](https://msdn.microsoft.com/en-us/library/windows/desktop/ms646312(v=vs.85).aspx) and the hwnd retrieved earlier.

Not easy from JavaScript but could be possible using [node-ffi](https://github.com/node-ffi/node-ffi).

But ideally this will be fixed in Electron itself.

**Comment by arieljannai** [comment]
I agree it should be from their side, but meanwhile, I use [this script](https://gist.github.com/arieljannai/81aef1937e5a832986c2cc56eab5550d#file-restore_focus_zazu_closed-ahk)

**Comment by bayleedev** [comment]
Thank you for the script! Is there a way to integrate that into zazu?

**Comment by arieljannai** [comment]
I'm not sure it would be a nice integration - maybe it can be suggested as a fix-hack to people who uses AutoHotKey?
If the user already has AHK installed, that won't be a problem - just starting the script (if not already running) when Zazu starts.

P.s the script is not perfect, I need to change it:
* The problems it has:
    * If the last active window was desktop - it will bring back the one before desktop
    * **[SOLVED]** The matching is not effective, since it's by window title (so also this chrome tab in Zazu project activates the script) - I need to change it to another matching - or maybe a combination of multiple criterias.

**Comment by arieljannai** [comment]
I think I solved the second problem in my script - I changed the filtering of the window to match by id. The only problem is that I don't know if the id is constant in other machines.

Could you check if the updated script works for you?
Or in addition, create a script with that code, activate Zazu and press `Alt+a` to get a popup with the `ahk_id` (so I'll know if it's the same or not)

```ahk
!A:: 
WinGet, WinID, ID, A
MyWin = ahk_id %WinID%
MsgBox, %MyWin%
Return
```

**Comment by arieljannai** [comment]
NVM, Changed it again.
Now it filters by the exe name, so it works

**Comment by Nantris** [comment]
@arieljannai Is this something that could be modified to work with other Electron apps?  I'd love to solve this problem myself, but I don't understand your solution. Is this something that could potentially be packaged into an Electron installer?

**Comment by arieljannai** [comment]
@Slapbox It could be integrated, but I wouldn't. At least not with my solution. It's simple and has bugs, but basically working. Still, it's a kind of workaround/patch rather than a solution for the problem.
If you would improve the script so it wouldn't have bugs - it could be integrated better.

**Comment by Nantris** [comment]
Ah I'd love to but I don't even understand what your snippet does. I will revisit this thread when the time comes though and see what I can do. I'd happily share a solution to this very frustrating problem if I could find one.

What sorts of bugs do you see with it? How exactly are you using it in conjunction with Zazu?
